### ISSUE TITLE
Getting parsed_args of subparsers

### ISSUE THREAD

**Issue by devnoop** [kislyuk/argcomplete#114]
following example:

```python
import argparse
import argcomplete
from argcomplete import warn

def MyCompleter(prefix, parsed_args, **kwargs):
    warn(parsed_args)
    return ['1','2','3','4']

parser = argparse.ArgumentParser()
parser.add_argument('--parser')
subparsers = parser.add_subparsers(dest='cmd')
sub = subparsers.add_parser('sub',help='initial project setup')
sub.add_argument('--subvar1',dest='sub1')
sub.add_argument('--subvar2',dest='sub2').completer = MyCompleter
argcomplete.autocomplete(parser)
args = vars(parser.parse_args())
```

`MyCompleter` only gets the parsed_args from parser but not from the subparser.  I didn't find any solution.

**Comment by jdappel** [comment]
I am dealing with this issue as well.

**Comment by redstreet** [comment]
I have the same problem. The issue might be upstream. @devnoop, your example works fine with Python 3.3.6, but fails to get parsed_args from the subparser in Python 3.5.1.

Python 3.3.6:
````
$ ./test.py sub --subvar1 ase --subvar2 
Namespace(cmd='sub', parser=None, sub1='ase', sub2=None)
````

Python 3.5.1:
````
$ ./test.py sub --subvar1 ase --subvar2 
Namespace(cmd='sub', parser=None)
````
I'm using argcomplete 1.0.0. @devnoop or @jdappel, does this test case look correct to you?

**Comment by jeromekelleher** [comment]
I'm hitting this issue also on Python 2.7. The problems seems to happen when argparse has an error in parsing the arguments to the subcommand's option. If there is an error, then the ``parsed_args`` is empty, and we don't get the values we need. We get an error in this case because the option expects exactly on argument, which it doesn't get. We can work around this by making the argument accept zero arguments with ``nargs='*'``, e.g.
```python
#!/usr/bin/env python
# PYTHON_ARGCOMPLETE_OK

import argparse
import argcomplete
from argcomplete import warn

def MyCompleter(prefix, parsed_args, **kwargs):
    warn(parsed_args)
    return ['1','2','3','4']

parser = argparse.ArgumentParser()
parser.add_argument('--parser')
subparsers = parser.add_subparsers(dest='cmd')
sub = subparsers.add_parser('sub',help='initial project setup')
sub.add_argument('--subvar1',dest='sub1')
sub.add_argument('--subvar2',dest='sub2', nargs="*").completer = MyCompleter
argcomplete.autocomplete(parser)
args = vars(parser.parse_args())
```

Then, we get 
```bash
./example.py sub --subvar1 x --subvar2 
 Namespace(cmd='sub', parser=None, sub1='x', sub2=[])
```
which has the value of sub1 populated.

Obviously we don't want to have to use nargs=* all the time; is a proper fix possible?

Oh yeah, I forgot to say; argcomplete is AWESOME!!!

**Comment by kislyuk** [comment]
Thanks @jeromekelleher!

The fundamental issue is that argcomplete internally updates the namespace atomically for the whole parse, not action-by-action. i.e. the equivalent of this:

```
def parse_known_args(...):
    ns = Namespace(cmd="foo")
    ns = self._do_parsing(ns)
    return ns
```

To fix this, we'll need to monkey-patch argparse some more in https://github.com/kislyuk/argcomplete/blob/master/argcomplete/my_argparse.py#L28.

**Comment by jeromekelleher** [comment]
> My goal in general is to avoid monkey-patching if at all possible (for obvious reasons of introducing brittleness into the system that's not relying on a well-defined API) but this seems to be a commonly requested improvement, so we can see if this can be done defensively. 

This is definitely the right approach @kislyuk, and it would be really great if we could find a solution to this issue. I would love to add argcomplete into a complicated CLI I'm working on, and this particular issue makes it not-quite-as-totally-amazing as it would be.

Thanks for the explanation and the pointers. I'm happy to do some work here if you can point me in the right direction. What I don't quite get yet is why this only happens when there are subcommands. We can have errors happening when there is an action at the top level (like in the GitHub example in the docs) and it's fine (the rest of the namespace is populated), but when there's an error within a subcommand, the namespace gets nuked. Why is this?

**Comment by kislyuk** [comment]
@jeromekelleher, I think I got this fixed. Released in v1.2.0, please test. Details in https://github.com/kislyuk/argcomplete/commit/bbaadbbe594c02ee9d2eff3b75653434e3a5f613.

**Comment by jeromekelleher** [comment]
It works! Thanks a million @kislyuk, this is great.
